Update docs for 2.0 (#352)
sunpy/ndcube: docs/data_classes.rst

.. _data_classes:

============

> Data Classes

Reviewer: I worry a little about the use of "data classes" here given that [`dataclasses`](https://docs.python.org/3/library/dataclasses.html) are now a thing in the standard library.

Should we just call them "classes", "objects" or "types"?

Author: I think `Data Objects` is fine.  I wanted to use `Data` to distinguish `NDCube`, `NDCubeSequence` and `NDCollection` from the coordinate objects like `Global Coords`, `ExtraCoords` and other used by `ndcube`.  Another option could be `ND Objects`.  So those three are members of the ND family.

Reviewer 2: I think `Data Objects` sounds good

============

> ndcube provides its features via its data classes: `~ndcube.NDCube`, `~ndcube.NDCubeSequence` and `~ndcube.NDCollection`.  This section describes the purpose of each and how they are structured and instantiated.  To learn how to slicem visualize and perform coordinate transformations with these classes, see the :ref:`slicing`, :ref:`plotting` and :ref:`coordinates` sections.

Reviewer 2: ```suggestion
ndcube provides its features via its data classes: `~ndcube.NDCube`, `~ndcube.NDCubeSequence` and `~ndcube.NDCollection`.  This section describes the purpose of each and how they are structured and instantiated.  To learn how to slice, visualize, and perform coordinate transformations with these classes, see the :ref:`slicing`, :ref:`plotting` and :ref:`coordinates` sections.
```

.. _ndcube:

NDCube
======

> ndcube's primary data class is `~ndcube.NDCube`.  It's designed for managing a single data array and set of WCS transformations.  `~ndcube.NDCube` provides unified slicing, visualization, coordinate conversion APIs as well as APIs for inspecting the data, coordinate transformations and metadata. `~ndcube.NDCube` does this in a way that is not specific to any number or physical type of axis.  It can therefore be used for any type of data (e.g. images, spectra, timeseries, etc.) so long as those data are represented by an array and a set of WCS transformations. Moreover, `~ndcube.NDCube` is agnostic to the fundamental array type in which the data is stored, as long as it behaves like a numpy array.

Reviewer 2: if data objects then reword here

```suggestion
ndcube's primary data object is `~ndcube.NDCube`.  It's designed for managing a single data array and set of WCS transformations.  `~ndcube.NDCube` provides unified slicing, visualization, coordinate conversion APIs as well as APIs for inspecting the data, coordinate transformations and metadata. `~ndcube.NDCube` does this in a way that is not specific to any number or physical type of axis.  It can therefore be used for any type of data (e.g. images, spectra, timeseries, etc.) so long as those data are represented by an array and a set of WCS transformations. Moreover, `~ndcube.NDCube` is agnostic to the fundamental array type in which the data is stored, as long as it behaves like a numpy array.
```

> Meanwhile, the WCS object can be any class, as long as it adhere's to the AstroPy `wcsapi (APE 14) <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_ specification. These features and flexibility make `~ndcube.NDCube` ideal for subclassing when creating tools for specific data types of data.  It enables developers and scientists to focus on developing the tools needed for their specific research while leveraging standarized APIs for non-data-type-specific functionalities (e.g. slicing).

Reviewer 2: ```suggestion
Meanwhile, the WCS object can be any class, as long as it adhere's to the AstroPy `wcsapi (APE 14) <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_ specification. These features and flexibility make `~ndcube.NDCube` ideal for subclassing when creating tools for specific data types of data.  It enables developers and scientists to focus on developing the tools needed for their specific research while leveraging standarized APIs for generic array functionalities (e.g. slicing).
```

Initialize an NDCube
--------------------
To initialize the most basic `~ndcube.NDCube` object, we need is a `numpy.ndarray`-like array containing the data and an APE-14-compliant WCS object (e.g. `astropy.wcs.WCS`) describing the coordinate transformations to and from array-elements. Let's create a 3-D array of data with shape ``(3, 4, 5)`` where every value is 1.

.. code-block:: python

  >>> import numpy as np
  >>> data = np.ones((3, 4, 5))

> Now let's create an `astropy.wcs.WCS` object.  Let the first world axis be wavelength, the second be helioprojective longitude, the third be helioprojective latitude. Remember that due to convention, the order WCS axes is reversed relative to the data array.  This means that the first two array axes will correspond to helioprojective latitude and longitude and the third to wavelength.

Reviewer 2: I think it would be good to give a real world example of what this would be. For example, this could be a data cube which represents 5 images at 5 different wavelengths lalalala...

.. code-block:: python

  >>> import astropy.wcs
  >>> wcs_input_dict = {
  ... 'CTYPE1': 'WAVE    ', 'CUNIT1': 'Angstrom', 'CDELT1': 0.2, 'CRPIX1': 0, 'CRVAL1': 10, 'NAXIS1': 5,
  ... 'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 0.5, 'CRPIX2': 2, 'CRVAL2': 0.5, 'NAXIS2': 4,

>   ... 'CTYPE3': 'HPLN-TAN', 'CUNIT3': 'deg', 'CDELT3': 0.4, 'CRPIX3': 2, 'CRVAL3': 1, 'NAXIS3': 3}

Reviewer: Would it not be cleaner to use the `WCS` API for this rather than constructing a dict?

Author: This is in API I always use.  Feel free to suggest an alternative.

Reviewer: ```python
wcs = WCS(naxis=3)
wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
...
```
etc.

  >>> input_wcs = astropy.wcs.WCS(wcs_input_dict)

Now we can create an `~ndcube.NDCube`.

.. code-block:: python

  >>> from ndcube import NDCube
  >>> my_cube = NDCube(data, input_wcs)

> The data array is stored in ``mycube.data`` while the WCS object is stored in ``my_cube.wcs``.  However, when manipulating/slicing the data it is better to slice the object as a whole.  (See section on :ref:`ndcube_slicing`.)  So the ``.data`` attribute should only be used to access specific raw data values.

Reviewer 2: I don't know why but :ref:`ndcube_slicing` isn't rendering a link on readthedocs

axis types, `~ndcube.NDCube.dimensions` and `~ndcube.NDCube.array_axis_physical_types`.

.. code-block:: python

  >>> my_cube.dimensions
  <Quantity [3., 4., 5.] pix>
  >>> my_cube.array_axis_physical_types
  [('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('em.wl',)]

`~ndcube.NDCube.dimensions` returns an `~astropy.units.Quantity` of pixel units giving the length of each dimension in the `~ndcube.NDCube` while `~ndcube.NDCube.array_axis_physical_types` returns tuples of strings denoting the types of physical properties represented by each array axis.  The tuples are arranged in array axis order.  As more than one physical type can be associated with an array axis, the length of each tuple can be greater than 1.  This is the case for the 1st and 2nd array array axes which are associated with the coupled world axes of helioprojective latitude and longitude. The axis names are in accordance with the International Virtual Observatory Alliance (IVOA)
`UCD1+ controlled vocabulary <http://www.ivoa.net/documents/REC/UCD/UCDlist-20070402.html>`_.

> `~ndcube.NDCube` provides many helpful features, specifically regarding coordinate transformations, slicing and visualization.  See the :ref:`cube_coordinates`, :ref:`cube_slicing` and :ref:`cube_plotting` sections.

Reviewer 2: similarly here the `cube_slicing` and `cube_plotting` not linking correctly on readthedocs

.. _ndcubesequence:

NDCubeSequence
==============

> `~ndcube.NDCubeSequence` is a class for handling multiple `~ndcube.NDCube` objects as if they were one contiguous data set.  The `~ndcube.NDCube` objects within an `~ndcube.NDCubeSequence` must be have the same shape and physical types associated with each axis.  They must also be arranged in some order.  The direction in which the cubes are ordered is referred to as the "sequence axis".  The sequence axis can be in addition and perpendicular to the cubes' axes, or parallel to one, i.e. they can be arranged in order along one of the cubes' axes.  For example, say we have two images with a shape of 512 x 512 represented by two 2-D `~ndcube.NDCube` objects.  Let's also say they that were taken at two different times, but that their WCS transformations only describe their celestial coordinates. We can place these `~ndcube.NDCube` objects into a `~ndcube.NDCubeSequence` where the sequence axis acts as a 3rd axis representing time.  Thus, the data set has an effective shape of ``(2, 512, 512)``.  However, say those two images are of adjacent regions of the sky.  In that case there is an ordering in one of the spatial axes, say the first.  If we set the common axis to that cube axis (``common_axis=0``), then we can also treat the data as if it were a single cube with a shape of ``(1024, 512)``.

Reviewer 2: ```suggestion
`~ndcube.NDCubeSequence` is a class for handling multiple `~ndcube.NDCube` objects as if they were one contiguous data set.  The `~ndcube.NDCube` objects within an `~ndcube.NDCubeSequence` must have the same shape and physical types associated with each axis.  They must also be arranged in some order.  The direction in which the cubes are ordered is referred to as the "sequence axis".  The sequence axis can be in addition and perpendicular to the cubes' axes, or parallel to one, i.e. they can be arranged in order along one of the cubes' axes.  For example, say we have two images with a shape of 512 x 512 represented by two 2-D `~ndcube.NDCube` objects.  Let's also say that they were taken at two different times, but that their WCS transformations only describe their celestial coordinates. We can place these `~ndcube.NDCube` objects into a `~ndcube.NDCubeSequence` where the sequence axis acts as a 3rd axis representing time.  Thus, the data set has an effective shape of ``(2, 512, 512)``.  However, say those two images are of adjacent regions of the sky.  In that case there is an ordering in one of the spatial axes, say the first.  If we set the common axis to that cube axis (``common_axis=0``), then we can also treat the data as if it were a single cube with a shape of ``(1024, 512)``.
```

Reviewer 2: I think an sketch or schematic here would be really helpful

Initializing an NDCubeSequence
------------------------------
To initialize the most basic `~ndcube.NDCubeSequence` object, all you need is a list of `~ndcube.NDCube` instances.  Let's first define three 3-D NDCubes for slit-spectrograph data as we did in the :ref:`ndcube` section of this tutorial.

.. code-block:: python

  >>> # Define data for cubes
  >>> import numpy as np
  >>> data0 = np.ones((3, 4, 5))
  >>> data1 = data0 * 2
  >>> data2 = data1 * 2

  >>> # Define WCS object for all cubes.
  >>> import astropy.wcs
  >>> wcs_input_dict = {
  ... 'CTYPE1': 'WAVE    ', 'CUNIT1': 'Angstrom', 'CDELT1': 0.2, 'CRPIX1': 0, 'CRVAL1': 10, 'NAXIS1': 5,
  ... 'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 0.5, 'CRPIX2': 2, 'CRVAL2': 0.5, 'NAXIS2': 4,

>   ... 'CTYPE3': 'HPLN-TAN', 'CUNIT3': 'deg', 'CDELT3': 0.4, 'CRPIX3': 2, 'CRVAL3': 1, 'NAXIS3': 3}

Reviewer: same comment as above.

`ndcube.NDCube.array_axis_physical_types`.  The sequence axis is given the label ``'meta.obs.sequence'`` as it is the IVOA UCD1+ controlled word that best describes it.  To call, simply do::

  >>> my_sequence.array_axis_physical_types
  [('meta.obs.sequence',),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'), ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('em.wl',)]

Once again, we can see the physical types associated with each axis in the cube-like paradigm be calling `ndcube.NDCubeSequence.cube_like_array_axis_physical_types` .

.. code-block:: python

  >>> my_sequence.cube_like_array_axis_physical_types
  [('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('em.wl',)]

Explode Along Axis
------------------
During analysis of some data - say of a stack of images - it may be
necessary to make some different fine-pointing adjustments to each

> image that isn't accounted for the in the original WCS translations,

Reviewer: While this is certainty the point of this method, I do worry about how easy this would be *in general* with APE 14. We should at least test this works with FITS-WCS

e.g. due to satellite wobble.  If these changes are not describable
with a single WCS object, it may be desirable to break up the N-D
sub-cubes of an `~ndcube.NDCubeSequence` into an sequence of sub-cubes
with dimension N-1. This would enable a separate WCS object to be
associated with each image and hence allow individual pointing

> adjustments.

Reviewer 2: again I think a schematic or something here would be very helpful , or a more in-depth high-level example- its a little bit hard to visualize

provides a useful method,
`~ndcube.NDCubeSequence.explode_along_axis`. To call it, simply provide
the number of the data cube axis along which you wish to break up the sub-cubes.

.. code-block:: python

  >>> exploded_sequence = my_sequence.explode_along_axis(0)

Assuming we are using the same ``my_sequence`` as above, with dimensions of
``(<Quantity 3.0 pix>, <Quantity 3.0 pix>, <Quantity 4.0 pix>, <Quantity 5.0 pix>)``, the ``exploded_sequence`` will be an `~ndcube.NDCubeSequence` of nine 2-D NDCubes each with shape ``(<Quantity 4.0 pix>, <Quantity 5.0 pix>)``.

.. code-block:: python

  >>> # Check old and new shapes of the squence
  >>> my_sequence.dimensions
  (<Quantity 3. pix>, <Quantity 3. pix>, <Quantity 4. pix>, <Quantity 5. pix>)
  >>> exploded_sequence.dimensions
  (<Quantity 9. pix>, <Quantity 4. pix>, <Quantity 5. pix>)

Note that an `~ndcube.NDCubeSequence` can be exploded along any axis.  A common axis need not be defined.

To learn how to slice `~ndcube.NDCubeSequence` instances and manipulate sequence coordinates, the :ref:`sequence_slicing` and :ref:`sequence_coordinates` sections.

.. _ndcollection:

NDCollection
============

> `~ndcube.NDCollection` is a container class for grouping `~ndcube.NDCube` or `~ndcube.NDCubeSequence` instances in an unordered way like in a Python `dict`.  `~ndcube.NDCollection` therefore is differs from `~ndcube.NDCubeSequence` in that the objects contained are not considered to be in any order, are not assumed to represent measurements of the same physical property, and they can have different dimensionalities.  However `~ndcube.NDCollection` is more powerful than a simple `dict` because it enables us to identify axes that are aligned between the objects and hence provides some limited slicing functionality.  (See :ref:`collection_slicing` to for more on slicing.)

Reviewer: Python dicts are kina ordered these days.

> `~ndcube.NDCollection` is a container class for grouping `~ndcube.NDCube` or `~ndcube.NDCubeSequence` instances in an unordered way like in a Python `dict`.  `~ndcube.NDCollection` therefore is differs from `~ndcube.NDCubeSequence` in that the objects contained are not considered to be in any order, are not assumed to represent measurements of the same physical property, and they can have different dimensionalities.  However `~ndcube.NDCollection` is more powerful than a simple `dict` because it enables us to identify axes that are aligned between the objects and hence provides some limited slicing functionality.  (See :ref:`collection_slicing` to for more on slicing.)

Reviewer: One line per sentence please.

> One possible application of `~ndcube.NDCollection` is linking observations with derived data products.  Let's say we have a 3D `~ndcube.NDCube` representing space-space-wavelength.  Then let's say we fit a spectral line in each pixel's spectrum and extract its linewidth.  Now we have a 2D spatial map of linewidth with the same spatial axes as the original 3D cube.  There is a clear relationship between these two objects and so it makes sense to store them together.  An `~ndcube.NDCubeSequence` is not appropriate here as the physical properties represented by the two objects is different, they do not have an order within their common coordinate space, and they do not have the same dimensionality.  Instead let's use an `~ndcube.NDCollection`.

Reviewer: One line per sentence

Let's use ``my_cube`` defined above as our observations cube and define a "linewidth cube".

.. code-block:: python

  >>> # Define derived linewidth NDCube
  >>> linewidth_data = np.ones((3, 4)) / 2 # dummy data
  >>> linewidth_wcs_dict = {
  ...    'CTYPE1': 'HPLT-TAN', 'CUNIT1': 'deg', 'CDELT1': 0.5, 'CRPIX1': 2, 'CRVAL1': 0.5, 'NAXIS1': 20,
  ...    'CTYPE2': 'HPLN-TAN', 'CUNIT2': 'deg', 'CDELT2': 0.4, 'CRPIX2': 2, 'CRVAL2': 1, 'NAXIS2': 10}
  >>> linewidth_wcs = astropy.wcs.WCS(linewidth_wcs_dict)
  >>> linewidth_cube = NDCube(linewidth_data, linewidth_wcs)

> To combine these ND objects into an `~ndcube.NDCollection`, simply supply a sequence of

Reviewer 2: ```suggestion
To combine these NDCube objects into an `~ndcube.NDCollection`, simply supply a sequence of
```

.. code-block:: python

  >>> my_collection.aligned_dimensions
  <Quantity [3., 4.] pix>

Note that this only tells us the lengths of the aligned axes.  To see the lengths of the non-aligned axes, e.g. the spectral axis of the ``observations`` object, you must inspect that ND object individually.

We can also see the physical properties to which the aligned axes correspond by using the `~ndcube.NDCollection.aligned_axis_physical_types` property.

.. code-block:: python

  >>> my_collection.aligned_axis_physical_types  # doctest: +SKIP
  [('custom:pos.helioprojective.lon', 'custom:pos.helioprojective.lat'), ('custom:pos.helioprojective.lon', 'custom:pos.helioprojective.lat')]

This returns the a `list` of `tuple` giving the physical types that correspond to each aligned axis.  For each aligned axis, only physical types are associated with all the cubes in the collection are returned.  Note that there is no there is no requirement that all aligned axes must represent the same physical types.  They just have to be the same length.  Therefore, is it possible that this property returns no physical types.

> The real power behing `~ndcube.NDCollection.aligned_axes` is that is enables all objects within the `~ndcube.NDCollection` to be sliced along the aligned axes simultaneously form the `~ndcube.NDCollection` level.  This allows users to quickly and accurately crop their entire data set to a region of interest, thereby speeding up their analysis workflow.  See the :ref:`collection_slicing` to see this in action.

Reviewer: provide an example of this?

Author: Examples are shown in the slicing section of the docs.

Reviewer: I did miss that, but I do wonder if it's worth a very terse example in line?

> The real power behing `~ndcube.NDCollection.aligned_axes` is that is enables all objects within the `~ndcube.NDCollection` to be sliced along the aligned axes simultaneously form the `~ndcube.NDCollection` level.  This allows users to quickly and accurately crop their entire data set to a region of interest, thereby speeding up their analysis workflow.  See the :ref:`collection_slicing` to see this in action.

Reviewer 2: ```suggestion
The real power behing `~ndcube.NDCollection.aligned_axes` is that it enables all objects within the `~ndcube.NDCollection` to be sliced along the aligned axes simultaneously from the `~ndcube.NDCollection` level.  This allows users to quickly and accurately crop their entire data set to a region of interest, thereby speeding up their analysis workflow.  See the :ref:`collection_slicing` to see this in action.
```

.. _data_classes:

==========

> ND Objects

Author: Suggest an alternative?

> ND Objects

Author: Going to leave this as is for this PR.  It certainly be changed in another PR if we come up with a better name.

==========
ndcube provides its features via its data objects: `~ndcube.NDCube`, `~ndcube.NDCubeSequence` and `~ndcube.NDCollection`.
This section describes the purpose of each and how they are structured and instantiated.
To learn how to slice, visualize, and perform coordinate transformations with these classes, see the :ref:`slicing`, :ref:`plotting` and :ref:`coordinates` sections.

.. _ndcube:

NDCube
======
`~ndcube.NDCube` is the primary data class the ndcube package.
It's designed to manage a single data array and set of WCS transformations.
`~ndcube.NDCube` provides unified slicing, visualization, coordinate transformation and self-inspection APIs which are independent of the number and physical types of axes.
It can therefore be used for any type of data (e.g. images, spectra, timeseries, etc.) so long as those data are represented by an object that behaves like a `numpy.ndarray` and the coordinates by an object that adheres to the AstroPy `WCS API (APE 14) <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_.

> This makes `~ndcube.NDCube` an powerful base class when developing tools for specific data types.

Reviewer 2: ```suggestion
This makes `~ndcube.NDCube` a powerful base class when developing tools for specific data types.
```

`~ndcube.NDCube` also provides classes for representing additional coordinates not included in the primary WCS object.
These are `~ndcube.ExtraCoords` (located at ``.extra_coords``) --- for additional coordinates associated with specific data axes --- and `~ndcube.GlobalCoords` (located at ``.global_coords``) for scalar coordinates associated with the `~ndcube.NDCube` as a whole.
These are discussed in the section in this guide on :ref:`coordinates`.

The figure below shows a schematic of an `~ndcube.NDCube` instance and the relationships between its components.
Array-based components are in blue (``.data``, ``.uncertainty``, and ``.mask``), metadata components in green (``.meta`` and ``.unit``), and coordinate components in red (``.wcs``, ``.extra_coords``, and ``.global_coords``).
Yellow ovals represent methods for inspecting, visualizing, and analyzing the `~ndcube.NDCube`.

.. image:: images/ndcube_diagram.png
  :width: 800
  :alt: Components of an NDCube

Initializing an NDCube
----------------------

> To initialize the most basic `~ndcube.NDCube` object, we need is a `numpy.ndarray`-like array containing the data and an APE-14-compliant WCS object (e.g. `astropy.wcs.WCS`) describing the coordinate transformations to and from array-elements.

Reviewer 2: ```suggestion
To initialize the most basic `~ndcube.NDCube` object, what we need is a `numpy.ndarray`-like array containing the data and an APE-14-compliant WCS object (e.g. `astropy.wcs.WCS`) describing the coordinate transformations to and from array-elements.
```

To instantiate a more complex `~ndcube.NDCube` with metadata, a data unit, uncertainties and a mask, we can  the following:

.. code-block:: python

  >>> import astropy.units as u
  >>> import astropy.wcs
  >>> import numpy as np
  >>> from astropy.nddata import StdDevUncertainty

  >>> from ndcube import NDCube

  >>> # Define data array.
  >>> data = np.random.rand(4, 4, 5)

  >>> # Define WCS transformations in an astropy WCS object.
  >>> wcs = astropy.wcs.WCS(naxis=3)
  >>> wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
  >>> wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
  >>> wcs.wcs.cdelt = 0.2, 0.5, 0.4
  >>> wcs.wcs.crpix = 0, 2, 2
  >>> wcs.wcs.crval = 10, 0.5, 1

  >>> # Define mask.  Initially set all elements unmasked.
  >>> mask = np.zeros_like(data, dtype=bool)
  >>> mask[0, 0][:] = True  # Now mask some values.
  >>> # Define uncertainty, metadata and unit.
  >>> uncertainty = StdDevUncertainty(np.sqrt(np.abs(data)))
  >>> meta = {"Description": "This is example NDCube metadata."}
  >>> unit = u.ct

  >>> # Instantiate NDCube with supporting data.
  >>> my_cube = NDCube(data, wcs=wcs, uncertainty=uncertainty, mask=mask, meta=meta, unit=unit)

> Generating `~ndcube.ExtraCoords` and `~ndcube.GlobalCoords` objects and attaching them to your `~ndcube.NDCube` is demonstrated in the :ref:`extra_coords` and :ref:`global_coords` sections.

Reviewer: This is certainly true for GC, less for EC I guess.

> Generating `~ndcube.ExtraCoords` and `~ndcube.GlobalCoords` objects and attaching them to your `~ndcube.NDCube` is demonstrated in the :ref:`extra_coords` and :ref:`global_coords` sections.

Author: This would be true for EC too did NDCube was instantiated with an empty EC object by default.  Unless there's an issue with empty EC coord objects or a performance consideration, we should instantiated NDCube with an empty EC object for consistency and ease-of-use.

.. image:: images/ndcubesequence_diagram.png
  :width: 800
  :alt: Schematic of an NDCubeSequence and its two configurations.

However, let's also say that the images represent tiles in a mosaic that, when combined, form a map of the sky much larger than the field of view of the instrument.
Thus the images represent adjacent regions of the sky.
In that case the cubes are not only ordered in time, but also along one of their spatial axes.
Another way of saying this is that the sequence axis is parallel to one of the cubes' axes.
This cube axis is known as the "common axis".
Let's say in our example that the common axis is the 2nd axis.
Thus, we can also treat the data set as if it were a single image with a shape of ``(512, 2048)``.
See panel b) of the figure above.

Setting a common axis is optional and does not force the user to interact with the data as if it were in configuration b).
Instead `~ndcube.NDCubeSequence` has different versions of its methods whose names are prefixed with ``cube_like`` that account for the common axis and equivalent non-cube-like methods that do not.

> This allows users to switch back and forth between configurations a) and b) as their use case demands without the requiring the user to have two versions of the same data.

Reviewer 2: ```suggestion
This allows users to switch back and forth between configurations a) and b) as their use case demands without the requiring of the user to have two versions of the same data.
```

This flexibility makes `~ndcube.NDCubeSequence` a powerful tool when handling complex N-D dimensional data described by different but comparable coordinate transformations.

Initializing an NDCubeSequence
------------------------------
To initialize the most basic `~ndcube.NDCubeSequence`, all you need is a list of `~ndcube.NDCube` instances.
Let's first define four 3-D NDCubes.

> .. code-block:: python

Author: Yes!

The sequence axis is given the label ``'meta.obs.sequence'`` as it is the IVOA UCD1+ controlled word that best describes it.
To call, simply do:

.. code-block:: python

  >>> my_sequence.array_axis_physical_types
  [('meta.obs.sequence',),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'), ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('em.wl',)]

Once again, we can see the physical types associated with each axis in the cube-like paradigm be calling `ndcube.NDCubeSequence.cube_like_array_axis_physical_types`.

.. code-block:: python

  >>> my_sequence.cube_like_array_axis_physical_types
  [('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('em.wl',)]

Explode Along Axis
------------------
During analysis of some data --- say of a stack of images --- it may be necessary to make some different fine-pointing adjustments to each image that isn't accounted for the in the original WCS translations, e.g. due to satellite wobble.

> If these changes are not describable with a single WCS object, it may be desirable to break up the N-D sub-cubes of an `~ndcube.NDCubeSequence` into an sequence of sub-cubes with dimension N-1.

Reviewer 2: ```suggestion
If these changes are not describable with a single WCS object, it may be desirable to break up the N-D sub-cubes of an `~ndcube.NDCubeSequence` into a sequence of sub-cubes with dimension N-1.
```

This would enable a separate WCS object to be associated with each image and hence allow individual pointing adjustments.

> Rather than manually dividing the cubes up and deriving the corresponding WCS object for each exposure, `~ndcube.NDCubeSequence` provides a useful method, `~ndcube.NDCubeSequence.explode_along_axis`.

Author: You're right!  Although `NDCubeSequence` has it too.

To call it, simply provide the number of the array axis along which you wish to break up the cubes.

.. code-block:: python

  >>> exploded_sequence = my_sequence.explode_along_axis(0)

Assuming we are using the same ``my_sequence`` as above, with dimensions of ``(<Quantity 4.0 pix>, <Quantity 4.0 pix>, <Quantity 4.0 pix>, <Quantity 5.0 pix>)``, the ``exploded_sequence`` will be an `~ndcube.NDCubeSequence` of nine 2-D NDCubes each with shape ``(4, 5)``.

.. code-block:: python

  >>> # Check old and new shapes of the squence
  >>> my_sequence.dimensions
  (<Quantity 4. pix>, <Quantity 4. pix>, <Quantity 4. pix>, <Quantity 5. pix>)
  >>> exploded_sequence.dimensions
  (<Quantity 16. pix>, <Quantity 4. pix>, <Quantity 5. pix>)

Note that an `~ndcube.NDCubeSequence` can be exploded along any axis.  A common axis need not be defined.

To learn how to slice `~ndcube.NDCubeSequence` instances and manipulate sequence coordinates, see the :ref:`sequence_slicing` and :ref:`sequence_coordinates` sections.

.. _ndcollection:

NDCollection
============
`~ndcube.NDCollection` is a container class for grouping `~ndcube.NDCube` or `~ndcube.NDCubeSequence` instances in an unordered way.

> `~ndcube.NDCollection` therefore is differs from `~ndcube.NDCubeSequence` in that the objects contained are not considered to be in any order, are not assumed to represent measurements of the same physical property, and they can have different dimensionalities.

Reviewer 2: ```suggestion
`~ndcube.NDCollection` therefore differs from `~ndcube.NDCubeSequence` in that the objects contained are not considered to be in any order, are not assumed to represent measurements of the same physical property, and they can have different dimensionalities.
```

.. code-block:: python

  >>> # Define derived linewidth NDCube
  >>> linewidth_data = np.random.rand(4, 4) / 2 # dummy data
  >>> linewidth_wcs = astropy.wcs.WCS(naxis=2)
  >>> linewidth_wcs.wcs.ctype = 'HPLT-TAN', 'HPLN-TAN'
  >>> linewidth_wcs.wcs.cunit = 'deg', 'deg'
  >>> linewidth_wcs.wcs.cdelt = 0.5, 0.4
  >>> linewidth_wcs.wcs.crpix = 2, 2
  >>> linewidth_wcs.wcs.crval = 0.5, 1
  >>> linewidth_cube = NDCube(linewidth_data, linewidth_wcs)

To combine these ND objects into an `~ndcube.NDCollection`, simply supply a sequence of ``(key, value)`` pairs in the same way that you initialize and dictionary.

.. code-block:: python

  >>> from ndcube import NDCollection
  >>> my_collection = NDCollection([("observations", my_cube), ("linewidths", linewidth_cube)])

To access each ND object in ``my_collection`` index it with the name of the desired object, just like a `dict`:

.. code-block:: python

  >>> my_collection["observations"]  # doctest: +SKIP

And just like a `dict` we can see the different names available using the ``keys`` method:

.. code-block:: python

  >>> my_collection.keys()
  dict_keys(['observations', 'linewidths'])

Aligned Axes
------------

> In the above example, the linewidth object's axes are aligned with the first two axes of observations object.

Reviewer 2: ```suggestion
In the above example, the linewidth object's axes are aligned with the first two axes of the observations object.
```

Designating these axes as aligned allows both members of the collection to be simultanouesly sliced, thus enabling users to quickly and accurately crop their entire data set to a region of interest.
(For more on this, see :ref:`collection_slicing`.)
There are a few ways to designate aligned axes.
If the members of the collection have the same axis ordering, as is the case in our example, we can provide a single `tuple` of `int`, designating the array axes that are aligned.
(Note that aligned axes must have the same lengths.)

.. code-block:: python

  >>> my_collection = NDCollection([("observations", my_cube), ("linewidths", linewidth_cube)],
  ...                              aligned_axes=(0, 1))

We can see which axes are aligned by inpecting the ``aligned_axes`` attribute:

.. code-block:: python

  >>> my_collection.aligned_axes
  {'observations': (0, 1), 'linewidths': (0, 1)}

This gives us the array axes for each ND object separately.

> We should read this as array axis 0 of ``observations`` is aligned with the array axis 0 of ``'linewidths'``, as so on.

Reviewer 2: ```suggestion
We should read this as array axis 0 of ``observations`` is aligned with the array axis 0 of ``'linewidths'``, and so on.
```

However, the mapping can be more complicated.

> Let's say we reversed the axes of our ``linewidths`` ND object for some reason:

Author: I think it's fine.

Because aligned axes must have the same lengths, we can get the lengths of the aligned axes by using the ``aligned_dimensions`` property.

.. code-block:: python

  >>> my_collection.aligned_dimensions
  <Quantity [4., 4.] pix>

Note that this only tells us the lengths of the aligned axes.
To see the lengths of the non-aligned axes, e.g. the spectral axis of the ``observations`` object, you must inspect that ND object individually.

We can also see the physical properties to which the aligned axes correspond by using the `~ndcube.NDCollection.aligned_axis_physical_types` property.

.. code-block:: python

  >>> my_collection.aligned_axis_physical_types  # doctest: +SKIP
  [('custom:pos.helioprojective.lon', 'custom:pos.helioprojective.lat'), ('custom:pos.helioprojective.lon', 'custom:pos.helioprojective.lat')]

This returns a `list` of `tuple` in array axis order giving the physical types that correspond to each aligned axis.
For each aligned axis, only physical types associated with all the cubes in the collection are returned.

> Note that there is no there is no requirement that all aligned axes must represent the same physical types.

Reviewer 2: ```suggestion
Note that there is no requirement that all aligned axes must represent the same physical types.
```

They just have to be the same length.

> Therefore, is it possible that this property returns no physical types.

Reviewer 2: ```suggestion
Therefore, it is possible that this property returns no physical types.
```
